Support git-rerere to record and replay conflict resolution

krobelus opened the pull request:
This adds basic support for [git-rerere].
Whenever a merge conflict in a file is successfully resolved by the user,
the conflict (the preimage) and the merged file (the postimage) are recorded.
When resolving the same conflict again, the recorded resolution is replayed,
unless it doesn't merge cleanly with the current preimage.

The resolutions are shared with the ones that git-rerere records;
both save them in .git/rr-cache/<conflict-id>.

My motivation is that I'm using git-revise in a script that sometimes needs
to resolve the same conflicts over and over again. By caching conflict
resolutions I only need to do that once.

This feature is disabled by default.  Use it by enabling revise.rerere (or
rerere.enabled) _and_ rerere.autoUpdate. In future, we plan to also support
use without rerere.autoUpdate.

Upstream git-rerere can remember multiple resolutions for each conflict,
so-called "variants".  This is not yet supported by this patch - we only
keep the last resolution for each conflict around. This should probably be
added in a future patch.

[git-rerere]: <https://git-scm.com/docs/git-rerere/>

alerque commented:
I've definitely missed rerere support in the past and will look forward to this landing.

mystor commented on gitrevise/merge.py:
```diff
@@ -199,29 +202,131 @@ def merge_blobs(
+    two = b""
+    context = hashlib.sha1()
+    for line in preimage.splitlines():
+        if line.startswith(b"<" * MARKER_SIZE):
+            hunk = RR_SIDE_1
+            continue
+        if line.startswith(b"|" * MARKER_SIZE):
+            if hunk != RR_SIDE_1:
+                break
+            hunk = RR_OTHER
+            continue
+        if line.startswith(b"=" * MARKER_SIZE):
+            hunk = RR_SIDE_2
+            continue
+        if line.startswith(b">" * MARKER_SIZE):
+            if one > two:
+                one, two = two, one
+            context.update(one + b"\0" + two + b"\0")
+            one = b""
+            two = b""
+        if hunk == RR_SIDE_1:
+            one += line + b"\n"
+            continue
+        if hunk == RR_OTHER:
+            continue
+        if hunk == RR_SIDE_2:
+            two += line + b"\n"
+            continue
+        assert False
```
This doesn't quite match the behaviour of git's hashing logic from `handle_path`. It doesn't handle edge cases like nested conflict markers exactly the same as the one used by git itself.

I think the easiest way to implement something like that would be to mimic the original structure as closely as possible. This would also provide the normalization logic required for writing out the recorded preimage, and performing merges. It'd probably look something like (again, largely untested):

```python
class ConflictParseFailed(Exception):
    pass


def normalize_conflict(
        lines: Iterator[bytes],
        marker_size: int,
        hasher: Optional["hashlib._Hash"],
) -> bytes:
    cur_hunk: Optional[bytes] = b""
    other_hunk: Optional[bytes] = None
    while True:
        line = next(lines, None)
        if line is None:
            raise ConflictParseFailed("unexpected eof")
        elif line.startswith(b"<" * marker_size):
            # parse recursive conflicts, including their processed output in the current hunk
            conflict = normalize_conflict(lines, marker_size, None)
            if cur_hunk is not None:
                cur_hunk += conflict
        elif line.startswith(b"|" * marker_size):
            # ignore the diff3 ||||||| section. Must be still parsing the first hunk.
            if other_hunk is not None:
                raise ConflictParseFailed("unexpected ||||||| conflict marker")
            (other_hunk, cur_hunk) = (cur_hunk, None)
        elif line.startswith(b"=" * marker_size):
            # switch into the second hunk after the ======= marker
            # could be in either the diff3 original section or the first hunk
            if cur_hunk is not None:
                if other_hunk is not None:
                    raise ConflictParseFailed(
                        "unexpected ======= conflict marker")
                other_hunk = cur_hunk
            cur_hunk = b""
        elif line.startswith(b">" * marker_size):
            # end of conflict. update hasher, and return a normalized conflict
            if cur_hunk is None or other_hunk is None:
                raise ConflictParseFailed("unexpected >>>>>>> conflict marker")

            hunk1, hunk2 = sorted([cur_hunk, other_hunk])
            if hasher:
                hasher.update(hunk1 + b"\0")
                hasher.update(hunk2 + b"\0")
            return b"".join((
                b"<" * marker_size + b"\n",
                hunk1,
                b"=" * marker_size + b"\n",
                hunk2,
                b">" * marker_size + b"\n",
            ))
        elif cur_hunk is not None:
            # add non-marker lines to the current hunk (or discard if in the diff3 hunk)
            cur_hunk += line


def normalize_conflicted_file(body: bytes,
                              marker_size: int = 7) -> Tuple[bytes, str]:
    hasher = hashlib.sha1()
    normalized = b""

    lines = iter(body.splitlines(keepends=True))
    while True:
        line = next(lines, None)
        print(f"line = {line}")
        if line is None:
            return normalized, hasher.hexdigest()
        elif line.startswith(b"<" * marker_size):
            normalized += normalize_conflict(lines, marker_size, hasher)
        else:
            normalized += line
```

mystor commented on gitrevise/merge.py:
```diff
@@ -199,29 +202,131 @@ def merge_blobs(
+            ),
+            current=Blob(repo, postimage),
+            base=Blob(repo, recorded_preimage),
+            other=Blob(repo, preimage),
```
You're going to need to normalize the conflict markers in your preimage both here, and before saving it to disk. That way we can make sure that they continue to match when pulling them back off disk.
https://github.com/git/git/blob/d166e6afe5f257217836ef24a73764eba390c58d/Documentation/technical/rerere.txt

mystor commented on gitrevise/merge.py:
```diff
@@ -199,29 +202,131 @@ def merge_blobs(
+    conflict_dir.parent.mkdir(exist_ok=True)
+    conflict_dir.mkdir(exist_ok=True)
+    print("Recording conflict resolution")
+    (conflict_dir / "preimage").write_bytes(preimage)
```
I think you need to normalize the conflict markers in this preimage before writing it to disk. 

It also _might_ be worth checking you're not overriding a file before writing it, but I imagine git will handle an overwritten preimage/postimage just fine.

Finally, we should probably catch exceptions thrown in this method, warn, and discard them, as we probably don't want to hard-fail if we fail to record a resolution.

krobelus replied:
Currently I'm still overwriting the files which I'd assume is not a problem.It seems better than not overwriting an old resolution which might no longer be usable because it doesn't merge cleanly.

I need to look into the Git implementation a bit more. I plan to implement variants as well, then there will no longer be the need to overwrite old conflict data.

mystor commented on gitrevise/merge.py:
```diff
@@ -199,29 +202,131 @@ def merge_blobs(
+    if conflict_id is None:
+        return
+
+    conflict_dir = repo.gitdir / "rr-cache" / conflict_id
+    conflict_dir.parent.mkdir(exist_ok=True)
+    conflict_dir.mkdir(exist_ok=True)
```
`conflict_dir.mkdir(exist_ok=True, parents=True)` should recursively make the directory for you.

mystor commented on gitrevise/merge.py:
```diff
@@ -199,29 +202,131 @@ def merge_blobs(
+
+def reuse_recorded_resolution(repo, preimage) -> Tuple[Optional[Blob], Optional[str]]:
+    rr_cache = repo.gitdir / "rr-cache"
+    # TODO Should we cache settings like this?
+    if not repo.bool_config("rerere.enabled", default=rr_cache.is_dir()):
+        return None, None
```
It might be good to also have a `revise.rerere` config value which we also check, and defaults to the standard rerere config value, so that if someone runs into an issue with git-revise's rerere implementation, they can turn that off without turning off rerere in general.

We can also consider caching this on the `Repository` in the future, but I don't think that's necessary in this initial PR.

mystor commented on tests/test_rerere.py:
```diff
@@ -0,0 +1,175 @@
+
+
+def test_conflict_id_by_file_contents():
+    two_conflicts = b"""\
+<<<<<<<
+a
+=======
+b
+>>>>>>>
+
+<<<<<<<
+c
+=======
+d
+>>>>>>>
+"""
```
Please include text after the various conflict markers here like would appear in an actual diff, like `<<<<<<< original thingamabob`

mystor commented on tests/test_rerere.py:
```diff
@@ -0,0 +1,175 @@
+"""
+        )
+    )
+
```
Would be nice to include a test for nested commit markers as well :-)

mystor commented on gitrevise/merge.py:
```diff
@@ -199,29 +202,131 @@ def merge_blobs(
+        return None, conflict_id
+    if not repo.bool_config("rerere.autoUpdate", default=False):
+        return None, conflict_id
+    postimage_path = conflict_dir / "postimage"
+    postimage = postimage_path.read_bytes()
+    recorded_preimage = (conflict_dir / "preimage").read_bytes()
```
These should probably be within a `try: except:` block which warns but does not throw, as we don't want to error out if we fail to read previous rerere conflict resolutions.

mystor reviewed (changes requested):
Thanks for writing this PR! This will be an awesome feature to support in git-revise.

I hadn't looked into the git rerere implementation before, and wanted to make sure I understood it before reviewing your patches. Feel free to let me know if I got anything wrong in my understanding while doing the review, I could've easily overlooked something.

krobelus commented:
Thanks for the detailed review! My initial implementation was a bit sloppy.
I have addressed all comments and added TODO notes for the follow-ups.
The tests are more useful now, I also test that recording a conflict with git and replaying it with git revise actually works.

I dropped the marker size since I'm not sure if it really adds value.

krobelus commented on gitrevise/merge.py:
```diff
@@ -199,29 +202,131 @@ def merge_blobs(
+    if not conflict_dir.is_dir():
+        return None, conflict_id
+    if not repo.bool_config("rerere.autoUpdate", default=False):
+        return None, conflict_id
```
Ok, I'll skip recording for now.

I agree, asking the user when `autoUpdate` is false seems like it corresponds best to what git does.

If we want to be able to override `autoUpdate`, we should probably turn `revise.rerere` into an enum, something like `always`/`ask`/`never`.

mystor commented on gitrevise/merge.py:
```diff
@@ -175,6 +178,67 @@ def merge_blobs(
+        return merged_blob
+
+    if input("  Edit conflicted file? (Y/n) ").lower() == "n":
+        raise MergeConflict("user aborted")  # pylint: disable=W0707
```
You shouldn't need the `pylint` comment here anymore, as this is no longer in an `except` block

mystor commented on gitrevise/merge.py:
```diff
@@ -175,6 +178,67 @@ def merge_blobs(
+
+    # Was the merge successful?
+    if input("  Merge successful? (y/N) ").lower() != "y":
+        raise MergeConflict("user aborted")  # pylint: disable=W0707
```
You shouldn't need the `pylint` comment here anymore

mystor commented on gitrevise/merge.py:
```diff
@@ -175,6 +178,67 @@ def merge_blobs(
+
+def merge_files(
+    labels: Tuple[str, str, str],
+    current: Blob,
+    base: Optional[Blob],
+    other: Blob,
```
nit: I think these three arguments would make more sense as `bytes` arguments, as we're currently being forced to create `Blob` objects from bytes in `replay_recorded_resolution` just to pass them to this function.

mystor reviewed (approved):
Thanks a ton for these patches, they look great! I've left a couple of comments, but I'll clean them up in a follow-up.
